File: src/06.hashMap/21.609.ts
export function findDuplicate(paths: string[]): string[][] {
    const map = new Map<string, string[]>();
    const ans: string[][] = [];

    for (const path of paths) {
        const files = path.split(" ");
        const dir = files[0];

        for (let i = 1; i < files.length; i++) {
            const str = /^(.+)\((.+)\)$/.exec(files[i]);
            if (!str) {
                continue;
            }

            const filename = str[1];
            const content = str[2];
            const p = dir + "/" + filename;

            if (map.has(content)) {
                map.get(content).push(p);
            } else {
                map.set(content, [p]);
            }
        }
    }

    for (const value of map.values()) {
        if (value.length >= 2) {
            ans.push(value);
        }
    }

    return ans;
}

File: src/06.hashMap/21.609.go
package hashmap

import (
	"strings"
)

func FindDuplicate(paths []string) [][]string {
	// ans := [][]string{}
	// mp := map[string][]string{}
	// reg := regexp.MustCompile(`^(.+)\((.+)\)$`)

	// for _, v := range paths {
	// 	files := strings.Split(v, " ")
	// 	dir := files[0]

	// 	for i := 1; i < len(files); i++ {
	// 		match := reg.FindStringSubmatch(files[i])

	// 		if match == nil {
	// 			continue
	// 		}

	// 		content := match[2]
	// 		path := dir + "/" + match[1]

	// 		mp[content] = append(mp[content], path)
	// 	}
	// }

	// for _, v := range mp {
	// 	if len(v) >= 2 {
	// 		ans = append(ans, v)
	// 	}
	// }

	// return ans

	// 使用 make 能减小内存, 不是 正则 减少时间
	ans := make([][]string, 0)
	mp := make(map[string][]string)

	for _, v := range paths {
		files := strings.Split(v, " ")
		dir := files[0]

		for i := 1; i < len(files); i++ {

			match := strings.Split(files[i], "(")

			content := match[1]

			mp[content] = append(mp[content], dir+"/"+match[0])
		}
	}

	for _, v := range mp {
		if len(v) >= 2 {
			ans = append(ans, v)
		}
	}

	return ans
}
